x/stake/keeper.go

```diff
@@ -0,0 +1,300 @@
+
+// get the most recently updated validators
+func (k Keeper) getAccUpdateValidators(ctx sdk.Context) (updates []Validator) {
+	store := ctx.KVStore(k.storeKey)
+
+	iterator := store.Iterator(subspace(AccUpdateValidatorsKey)) //smallest to largest
+	for ; iterator.Valid(); iterator.Next() {
+		valBytes := iterator.Value()
+		var val Validator
+		err := k.cdc.UnmarshalBinary(valBytes, &val)
+		if err != nil {
+			panic(err)
+		}
+		updates = append(updates, val)
+	}
+	iterator.Close()
+	return
+}
+
+// remove all validator update entries
+func (k Keeper) clearAccUpdateValidators(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	k.deleteSubSpace(store, AccUpdateValidatorsKey)
+}
+
+// TODO move to common functionality somewhere
+func (k Keeper) deleteSubSpace(store sdk.KVStore, key []byte) {
+	iterator := store.Iterator(subspace(key))
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+	iterator.Close()
+}
+
+//_____________________________________________________________________
+
+func (k Keeper) getDelegatorBond(ctx sdk.Context,
+	delegatorAddr, candidateAddr sdk.Address) (bond DelegatorBond, found bool) {
+
+	store := ctx.KVStore(k.storeKey)
+	delegatorBytes := store.Get(GetDelegatorBondKey(delegatorAddr, candidateAddr, k.cdc))
+	if delegatorBytes == nil {
+		return bond, false
+	}
+
+	err := k.cdc.UnmarshalBinary(delegatorBytes, &bond)
+	if err != nil {
+		panic(err)
+	}
+	return bond, true
+}
+
+// load all bonds of a delegator
+func (k Keeper) getDelegatorBonds(ctx sdk.Context, delegator sdk.Address, maxRetrieve int16) (bonds []DelegatorBond) {
+	store := ctx.KVStore(k.storeKey)
+	delegatorPrefixKey := GetDelegatorBondsKey(delegator, k.cdc)
+	iterator := store.Iterator(subspace(delegatorPrefixKey)) //smallest to largest
+
+	bonds = make([]DelegatorBond, maxRetrieve)
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() || i > int(maxRetrieve-1) {
+			iterator.Close()
+			break
+		}
+		bondBytes := iterator.Value()
+		var bond DelegatorBond
+		err := k.cdc.UnmarshalBinary(bondBytes, &bond)
+		if err != nil {
+			panic(err)
+		}
+		bonds[i] = bond
+		iterator.Next()
+	}
+	return bonds[:i] // trim
+}
+
+func (k Keeper) setDelegatorBond(ctx sdk.Context, bond DelegatorBond) {
+	store := ctx.KVStore(k.storeKey)
+	b, err := k.cdc.MarshalBinary(bond)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(GetDelegatorBondKey(bond.DelegatorAddr, bond.CandidateAddr, k.cdc), b)
+}
+
+func (k Keeper) removeDelegatorBond(ctx sdk.Context, bond DelegatorBond) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(GetDelegatorBondKey(bond.DelegatorAddr, bond.CandidateAddr, k.cdc))
+}
+
+//_______________________________________________________________________
+
+// load/save the global staking params
+func (k Keeper) GetParams(ctx sdk.Context) (params Params) {
+	// check if cached before anything
+	if k.params != (Params{}) {
+		return k.params
+	}
+	store := ctx.KVStore(k.storeKey)
+	b := store.Get(ParamKey)
+	if b == nil {
```

Comment:
Desirable to move store initialization logic into a single function for clarity? Also avoids strange results were a store to return nil incorrectly.

Reply:
created an issue, I think this can wait until after the MVP though https://github.com/cosmos/cosmos-sdk/issues/724